runtime/gc_modron_startup/mminit.cpp

```diff
@@ -3030,6 +3035,34 @@ gcInitializeDefaults(J9JavaVM* vm)
 	return J9VMDLLMAIN_FAILED;
 }
 
+void
+initializeIndexableObjectHeaderSizes(J9JavaVM* vm) {
+	vm->isDataAddrPresent = TRUE;
+
+	if (J9JAVAVM_COMPRESS_OBJECT_REFERENCES(vm)) {
+		vm->sizeOfJ9IndexableObjectContiguous = sizeof(J9IndexableObjectContiguousCompressed);
+		vm->sizeOfJ9IndexableObjectDiscontiguous = sizeof(J9IndexableObjectDiscontiguousCompressed);
+	} else {
+		vm->sizeOfJ9IndexableObjectContiguous = sizeof(J9IndexableObjectContiguousFull);
+		vm->sizeOfJ9IndexableObjectDiscontiguous = sizeof(J9IndexableObjectDiscontiguousFull);
+	}
+
+#if defined(J9VM_ENV_DATA64)
+	if (vm->isIndexableDualHeaderShapeEnabled && (J9_GC_POLICY_BALANCED != ((OMR_VM *)vm->omrVM)->gcPolicy)) {
+		vm->isDataAddrPresent = FALSE;
+		if (J9JAVAVM_COMPRESS_OBJECT_REFERENCES(vm)) {
+			vm->sizeOfJ9IndexableObjectContiguous = sizeof(J9IndexableObjectContiguousCompressed) - sizeof(void *);
+			vm->sizeOfJ9IndexableObjectDiscontiguous = sizeof(J9IndexableObjectDiscontiguousCompressed) - sizeof(void *);
+		} else {
+			vm->sizeOfJ9IndexableObjectContiguous = sizeof(J9IndexableObjectContiguousFull) - sizeof(void *);
+			vm->sizeOfJ9IndexableObjectDiscontiguous = sizeof(J9IndexableObjectDiscontiguousFull) - sizeof(void *);
+		}
```

Comment:
This code contains hidden assumption `dataAddr` field has `void *` type and it is last in the header. I am not sure how to articulate this fact better in the code. We need comment this fact at least.

Reply:
Right, we can do something like this
1) removing type assumption:
UDATA sizeOfDataAddrForFull = sizeof(((J9IndexableObjectContiguousFull *)NULL)->dataAddr);

2) asserting that it's the last field:
assert that sizeof(J9IndexableObjectDiscontiguousFull) == (offsetof(J9IndexableObjectContiguousFull, dataAddr) + sizeOfDataAddrForFull)

Reply:
We can also introduce new structs, but it will require more changes. For example:

```
typedef struct J9IndexableObjectContiguousFull {
	UDATA clazz;
	U_32 size;
#if defined(J9VM_ENV_DATA64)
	U_32 padding;
	void *dataAddr;
#endif /* J9VM_ENV_DATA64 */
} J9IndexableObjectContiguousFull;

typedef struct J9IndexableObjectWithDataAddrContiguousFull {
	J9IndexableObjectContiguousFull header;
#if defined(J9VM_ENV_DATA64)
	void *dataAddr;
#endif /* J9VM_ENV_DATA64 */
} J9IndexableObjectWithDataAddrContiguousFull;
```

Reply:
Yes, my initial thought was we should introduce new structs. And I still don't like the fact we are writing first object slot(s) over the object header formally. But as you said it requires more changes. Is it reasonable?

Reply:
> Not sure what you mean by 'writing first object slot(s) over the object header'.|

If indexable object header always (on 64-bit) contains `dataAddr` formally by disregarding it (use smaller size of the header) technically means first one or two slots of the object are written over place where `dataAddr` should virtually be according header definition.

However I agree formal definition of object headers is a virtual way to get access to object header fields (mapping only) using C/C++. With a new dynamic approach when compile time sizeof header can not be used any more we need to operate with header sizes stored in VM any way. The only problem we have left is control that access to `dataAddr` field is not accessed when not available (which in old static behaviour  is use to be solved by selecting applicable object header struct). Continuing to use selection of the right header struct might be not an optimum any more.
So, I think it is reasonable to have one general struct for the header (including `dataAddr`) for both cases as far as it is for "mapping" of header fields only. And we should have a mechanism to control `dataAddr` is not accessed when not available

Reply:
Ah, you meant data portion of objects slots.

It would be nice to control dataAddr field better (recognize if is attempted to access, if not effectively present), but it would be high runtime cost for real/production build. We can have a special debug build, though: make the field 'private', introduce get/set accessors and (in debug build only) assert that isDataAddrPresent it set.

I'm personally not that worried. The structs are now just a super sets of what the runtime header shape is. With a limited number of API/methods that really deal with the header shape, future bugs are not that likely.

Reply:
Sounds good. Marking this conversation as "resolved" for now as we have made the decision offline to introduce new structures. I will push these changes once complete and then we can have a new discussion around the updated changes.